### PR TITLE
planner/core: support named window

### PR DESCRIPTION
**Pull request by lamxTyler**
### What problem does this PR solve? 

We need to support the named window in the planner. Ref https://dev.mysql.com/doc/refman/8.0/en/window-functions-named-windows.html
### What is changed and how it works?

Before processing the window functions, we first resolve named window definitions and store it in the plan builder. After that, when processing the window functions, if it uses a named window, we will substitue the definitions for it.

### Check List 

Tests 

 - Unit test

Code changes

 - Has exported function/method change

Side effects

 - None

Related changes

 - None

---
This change is [Reviewable](https://reviewable.io/reviews/pingcap/tidb/8937)

### CODE REVIEW

File: planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -2678,6 +2693,68 @@ func (b *PlanBuilder) buildWindowFunction(p LogicalPlan, expr *ast.WindowFuncExp
...
+			name := spec.Name.O
+			if name == "" {
+				name = "<unnamed window>"
+			}
+			return ErrWindowNoRedefineOrderBy.GenWithStackByArgs(name, ref.Name.O)
+		}
+		spec.OrderBy = ref.OrderBy
+	}
+	if spec.PartitionBy != nil {
+		return ErrWindowNoChildPartitioning
```
Comment (winoros) [inline planner/core/logical_plan_builder.go]: Without `GenWithStackByArgs`, the error won't add error stask info. So you may need `errors.Trace()` here if you want the stack.

File: planner/core/logical_plan_builder.go [hunk 2]
```diff
@@ -2678,6 +2693,68 @@ func (b *PlanBuilder) buildWindowFunction(p LogicalPlan, expr *ast.WindowFuncExp
 	return window, nil
 }
 
+// resolveWindowSpec resolve window specifications for sql like `select ... from t window w1 as (w2), w2 as (partition by a)`.
+// We need to resolve the referenced window to get the definition of current window spec.
+func resolveWindowSpec(spec *ast.WindowSpec, specs map[string]ast.WindowSpec, inStack map[string]bool) error {
+	if inStack[spec.Name.L] {
+		return ErrWindowCircularityInWindowGraph
```
Comment (winoros) [inline planner/core/logical_plan_builder.go]: ditto

File: planner/core/logical_plan_builder.go [hunk 3]
```diff
@@ -2678,6 +2693,68 @@ func (b *PlanBuilder) buildWindowFunction(p LogicalPlan, expr *ast.WindowFuncExp
...
+			name := spec.Name.O
+			if name == "" {
+				name = "<unnamed window>"
+			}
+			return ErrWindowNoRedefineOrderBy.GenWithStackByArgs(name, ref.Name.O)
+		}
+		spec.OrderBy = ref.OrderBy
+	}
+	if spec.PartitionBy != nil {
+		return errors.Trace(ErrWindowNoChildPartitioning)
```
Comment (eurekaka) [inline planner/core/logical_plan_builder.go]: This seems incompatible with MySQL?  According to https://dev.mysql.com/doc/refman/8.0/en/window-functions-named-windows.html, `partition by` clause can be in referenced window while `over` clause can be without it?

Comment (eurekaka) [reply planner/core/logical_plan_builder.go]: According to the documentation, MySQL works for queries like:
```
OVER (w ORDER BY country)
... WINDOW w AS (PARTITION BY country)
```
in this case, `partition by` is defined in named window, so `spec.PartitionBy` should be `nil` now, we would raise error for this kind of query?

Comment (lamxTyler) [reply planner/core/logical_plan_builder.go]: We won't raise error since the check condition is `partitionBy != nil`?

Comment (eurekaka) [reply planner/core/logical_plan_builder.go]: Oh my mistake, then how about
```
OVER (w PARTITION BY country)
... WINDOW w AS (ORDER BY country)
```
would MySQL raise error? (I don't have MySQL 8.0 instance ...)

Comment (lamxTyler) [reply planner/core/logical_plan_builder.go]: Yes, MySQL would raise an error.

**Pull request merged**